components/fxa-client/src/internal/state_manager.rs

```diff
@@ -0,0 +1,446 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+};
+
+use crate::{
+    internal::{
+        commands, device,
+        oauth::{AccessTokenInfo, RefreshToken},
+        profile::Profile,
+        state_persistence::{state_from_json, state_to_json},
+        CachedResponse, Config, OAuthFlow, PersistedState,
+    },
+    AuthState, Error, EventListener, FxaEvent, Result, ScopedKey, StorageHandler,
+};
+
+/// Stores and manages the current state of the FxA client
+///
+/// All fields are private, which means that all state mutations must go through this module.  This
+/// makes it easier to reason about state changes.
+pub struct StateManager {
+    /// State that's persisted to disk
+    persisted_state: PersistedState,
+    /// In-progress OAuth flows
+    flow_store: HashMap<String, OAuthFlow>,
+    /// Current authorization state
+    auth_state: AuthState,
+    /// Application-supplied storage handlers
+    storage_handler: Option<Box<dyn StorageHandler>>,
+    /// Application-supplied event listener.  Wrapped with an Arc since it's shared with
+    /// FirefoxAccount
+    pub event_listener: Option<Arc<Box<dyn EventListener>>>,
+}
+
+impl StateManager {
+    /// Construct a new state manager
+    ///
+    /// The `uninitialized` flag signals that the application is not yet ready to perform disk IO
+    /// and `persisted_state` is a placeholder that may be replaced when `initialize()` is called.
+    pub(crate) fn new(persisted_state: PersistedState, uninitialized: bool) -> Self {
+        Self {
+            auth_state: Self::initial_auth_state(&persisted_state, uninitialized),
+            persisted_state,
+            flow_store: HashMap::new(),
+            storage_handler: None,
+            event_listener: None,
+        }
+    }
+
+    fn initial_auth_state(persisted_state: &PersistedState, uninitialized: bool) -> AuthState {
+        if uninitialized {
+            AuthState::Uninitialized
+        } else if persisted_state.session_token.is_some() || persisted_state.refresh_token.is_some()
+        {
+            AuthState::Connected
```

Comment:
As I mentioned on the call, it's not totally straightforward to know when an account is connected vs disconnected by looking at the persisted state.  But this test seems pretty reasonable to me.